test/test_array.fx

```diff
@@ -29,20 +29,52 @@ TEST("array.solve", fun() {
     EXPECT_NEAR(A*Ainv, I, 1e-10)
 })
 
-TEST("array.determinant", fun() {
-    val A =
-     [| 2., 1., 0., 4.;
-        0., 1., 1., 3.;
-        1., 0., 1., 2.;
-        2., 2., 0., 1. |]
-    val eps = 1e-10
-    val I = diag(4, 1.)
-    val As =
-     [| \A[0,:];
-        \A[1,:];
-        \(A[0,:] + A[1,:]);
-        \A[3,:] |]
-    EXPECT_NEAR(det(A), -15., eps)
-    EXPECT_NEAR(det(I), 1., eps)
-    EXPECT_NEAR(det(As), 0., eps)
+TEST("array.tuple_index", fun() {
+    val (h,w,d) = (10, 10, 10)
+    val rng = RNG(0xffffffffUL)
+    val A1 = random(rng, (h), -2., 2.)
+    val B1 = random(rng, (h), -2., 2.)
+    val A2 = random(rng, (h,w), -2., 2.)
+    val B2 = random(rng, (h,w), -2., 2.)
+    val A3 = random(rng, (h,w,d), -2., 2.)
+    val B3 = random(rng, (h,w,d), -2., 2.)
+
+    fun uniadd(A,B) { //TODO: Inefficient, must be implemented more precise. All of these functions.
+        var res = A
+        for x@idx <- A, y <- B { res[idx] = x + y}
+        res
+    }
+
+    fun add1(A,B) {
+        var res = A
+        for x@idx <- A, y <- B { 
+            val i:int = idx
+            res[i] = x + y}
+        res
+    }
+
+    fun add2(A,B) {
+        var res = A
+        for x@idx <- A, y <- B { 
+            val (i, j) = idx
+            res[i, j] = x + y}
+        res
+    }
+
+    fun add3(A,B) {
+        var res = A
+        for x@idx <- A, y <- B { 
+            val (i, j, k) = idx
+            res[i, j, k] = x + y}
+        res
+    }
+
+    EXPECT_EQ(uniadd(A1,B1), add1(A1,B1))
+    EXPECT_EQ(uniadd(A2,B2), add2(A2,B2))
+    val C3 = uniadd(A3,B3)
+    val C3r = add3(A3,B3)
+    for layer_num <- 0:h { //TODO: There is no Arr[,,] -> string conversion, so EXPECT_EQ cannot work with 3D arrays.
```

Comment:
insert the following string() function into Builtins.fx to make EXPECT_EQ(...3 ... 3 ... 3 ... 3 ...) compile:

```
fun string(a: 't [,,])
{
    val (d, m, n) = size(a)
    val planes = [| for k <- 0:d {
        val rows = [| for i <- 0:m {
            val elems = [| for j <- 0:n {repr(a[k, i, j])} |]
            join(", ", elems)
        } |]
        join_embrace("", "", ";\n", rows)
    } |]
    join_embrace("[| ", " |]", ";;\n", planes)
}
```

Reply:
Done.